Modules/_decimal/_decimal.c

```diff
@@ -3675,34 +3673,54 @@ dec_as_long(PyObject *dec, PyObject *context, int round)
     }
 
     status = 0;
-    ob_digit = NULL;
-#if PYLONG_BITS_IN_DIGIT == 30
-    n = mpd_qexport_u32(&ob_digit, 0, PyLong_BASE, x, &status);
-#elif PYLONG_BITS_IN_DIGIT == 15
-    n = mpd_qexport_u16(&ob_digit, 0, PyLong_BASE, x, &status);
-#else
-    #error "PYLONG_BITS_IN_DIGIT should be 15 or 30"
-#endif
+    int64_t val = mpd_qget_i64(x, &status);
 
-    if (n == SIZE_MAX) {
-        PyErr_NoMemory();
+    if (!status) {
+        mpd_del(x);
+        return PyLong_FromInt64(val);
+    }
+    assert(!mpd_iszero(x));
+
+    const PyLongLayout *layout = PyLong_GetNativeLayout();
+
+    assert(layout->bits_per_digit < 32);
+    assert(layout->digits_order == -1);
+    assert(layout->digit_endianness == (PY_LITTLE_ENDIAN ? -1 : 1));
+    assert(layout->digit_size == 2 || layout->digit_size == 4);
+
+    uint32_t base = (uint32_t)1 << layout->bits_per_digit;
+    size_t n, len = mpd_sizeinbase(x, base);
+    void *digits;
+    PyLongWriter *writer = PyLongWriter_Create(mpd_isnegative(x), len, &digits);
+
+    if (writer == NULL) {
         mpd_del(x);
         return NULL;
     }
 
-    if (n == 1) {
-        sdigit val = mpd_arith_sign(x) * ob_digit[0];
-        mpd_free(ob_digit);
+    status = 0;
+    /* mpd_qexport_*() functions used here with assumption, that no resizing
+       occur, i.e. len was obtained by a call to mpd_sizeinbase.  Set digits
+       to zero, as len can be overestimated. */
```

Comment:
```suggestion
    /* The mpd_qexport_*() functions used here assume that no resizing occurs,
       that is, 'len' was obtained via mpd_sizeinbase().  We also fill 'digits'
       with zeros first since 'len' can be overestimated. */
```

I used quotes just to indicate that those are variables and not just words.